### PR TITLE
Add dir() to ObjectProtocol.

### PR DESCRIPTION
**Pull request by m-ou-se**

### CODE REVIEW

**Comment by kngwyu** [comment]
@m-ou-se 
Could you please rebase on the master?
#888 would resolve the travis failure.
In addition, any test for this API is appriciated.

**Comment by davidhewitt** [comment]
A general comment about these kinds of functions - `dir`, `repr` etc.

At the moment they're trait methods, but I personally would prefer them to be free functions like this:

```
fn dir<T>(obj: T) -> &PyList
where T: AsRef<PyAny>
{ ... }

// Use later as 
pyo3::dir(obj);
```

It's not a _strong_ feeling, but I think it is more consistent with how these work in Python, where they're also used as `dir(obj)`, `repr(obj)` etc.

Wonder how you feel about this? 

(If we wanted compatibility we could leave them on `ObjectProtocol` but mark as deprecated?)

**Comment by kngwyu** [comment]
> A general comment about these kinds of functions - dir, repr etc.

A bit off-topic and should be discussed in another thread?

**Comment by davidhewitt** [comment]
> A bit off-topic and should be discussed in another thread?

I thought it was relevant given we're adding `dir` now, but sure I'll move this to another issue.

**Comment by davidhewitt** [comment]
Many thanks @m-ou-se for updating this PR. As I guess you have a need for `dir()`, I'd be interested to hear what you think of the idea proposed in #892 . (We could have `pyo3::dir()` and `o.__dir__()` in the way suggested by that PR.)

**Comment by m-ou-se** [comment]
@davidhewitt Ah, didn't see that one yet. Looks good.

**Comment by davidhewitt** [comment]
> 
> 
> @davidhewitt Ah, didn't see that one yet. Looks good.

I think I pushed that PR about 20 seconds after you updated this branch¬ 😄  I'll wait for some more feedback to see what others think of that PR, but I will definitely make sure to add `dir()` .

File: src/objectprotocol.rs [hunk 1]
```diff
@@ -545,4 +552,15 @@ mod test {
...
         assert_eq!(unsafe { obj.get_type().as_type_ptr() }, obj.get_type_ptr())
     }
+
+    #[test]
+    fn test_dir() {
+        let gil = Python::acquire_gil();
+        let py = gil.python();
+        let obj = py.eval("42", None, None).unwrap();
+        let dir = py.eval("dir(42)", None, None).unwrap().extract::<&PyList>().unwrap();
+        let a = obj.dir().into_iter().map(|x| x.extract::<String>().unwrap());
```
Comment (kngwyu) [inline src/objectprotocol.rs:562]: rustfmt complains about this

```suggestion
        let dir = py
            .eval("dir(42)", None, None)
            .unwrap()
            .extract::<&PyList>()
            .unwrap();
        let a = obj
            .dir()
            .into_iter()
            .map(|x| x.extract::<String>().unwrap());
```

**Comment by davidhewitt** [comment]
FWIW please do merge this PR without waiting for #892 - we might wait a little while to decide the right design there.

**Comment by kngwyu** [comment]
Thank you!

**Pull request merged**